### PR TITLE
Introducing composite max-size checker for cache

### PR DESCRIPTION
**Pull request by serkan-ozal**
Required for EE counterpart for this issue https://github.com/hazelcast/hazelcast-enterprise/issues/681

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/cache/impl/maxsize/impl/CompositeMaxSizeChecker.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+         */
+        OR;
+
+    }
+
+    public static CompositeMaxSizeChecker newCompositeMaxSizeChecker(CompositionOperator compositionOperator,
+                                                                     MaxSizeChecker... maxSizeCheckers) {
+        Preconditions.isNotNull(compositionOperator, "Composition operator cannot be null!");
+        Preconditions.isNotNull(maxSizeCheckers, "MaxSizeChecker's cannot be null!");
+        if (maxSizeCheckers.length == 0) {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/maxsize/impl/CompositeMaxSizeChecker.java]: is there a unit test for this branch?

Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/maxsize/impl/CompositeMaxSizeChecker.java]: no

**Comment by jerrinot** [comment]
:+1: (when the missing test is added:)

I am thinking whether we can establish a rule/guideline as: "a checker cheaper to test should be as a 1st/last/etc"

**Comment by serkan-ozal** [comment]
`MaxSizeChecker` is not in public API for eviction. In fact it is not in internal API for eviction. It is specific to JCache eviction. So no need a rule at the moment, not now

**Comment by jerrinot** [comment]
I am fine with that.

**Comment by mdogan** [comment]
:+1:

**Comment by ahmetmircik** [comment]
:+1:

**Comment by pveentjer** [comment]
Having an untested branch will reduce our test coverage (especially branch coverage). So if you don't add the test now, it needs to be added later.

**Comment by serkan-ozal** [comment]
@pveentjer I think you need some coffee. I had added them after comment. Just look at the PR and you will see the added tests for related branches

**Comment by pveentjer** [comment]
I think I had too much coffee since my wife is not at home and I normally am allowed to have only 2. Thanks for adding the test!

**Comment by jerrinot** [comment]
a test failure: https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-pr-builder/com.hazelcast$hazelcast/17726/testReport/junit/com.hazelcast.instance/OutOfMemoryErrorDispatcherTest/test_DefaultOutOfMemoryHandler_artificial_oome/

```

defaultOutOfMemoryHandler.onOutOfMemory(
    java.lang.OutOfMemoryError,
    [Mock for HazelcastInstance, hashCode: 627609502]
);
Never wanted here:
-> at com.hazelcast.instance.OutOfMemoryErrorDispatcherTest.test_DefaultOutOfMemoryHandler_artificial_oome(OutOfMemoryErrorDispatcherTest.java:155)
But invoked here:
-> at com.hazelcast.instance.OutOfMemoryErrorDispatcher.onOutOfMemory(OutOfMemoryErrorDispatcher.java:187)
```

**Comment by serkan-ozal** [comment]
run-lab-run

**Pull request merged**